tasks/ast_codegen/src/generators/visit.rs

```diff
@@ -0,0 +1,668 @@
+                let (args_def, args) = args
+                    .map(|it| it.parse_args_with(VisitArgs::parse))
+                    .map(|it| {
+                        it.into_iter()
+                            .flatten()
+                            .fold((Vec::new(), Vec::new()), Self::visit_args_fold)
+                    })
+                    .unwrap_or_default();
+                let visit = self.get_visitor(
+                    &typ,
+                    matches!(
+                        typ_wrapper,
+                        TypeWrapper::Vec | TypeWrapper::VecBox | TypeWrapper::OptVec
+                    ),
+                    visit_as.as_ref(),
+                );
+                let name = it.ident.as_ref().expect("expected named fields!");
+                let mut result = match typ_wrapper {
+                    TypeWrapper::Opt | TypeWrapper::OptBox | TypeWrapper::OptVec => quote! {
+                        if let Some(ref #name) = it.#name {
+                            visitor.#visit(#name #(#args)*);
+                        }
+                    },
+                    TypeWrapper::VecOpt => quote! {
+                        for #name in (&it.#name).into_iter().flatten() {
+                            visitor.#visit(#name #(#args)*);
+                        }
+                    },
+                    _ => quote! {
+                        visitor.#visit(&it.#name #(#args)*);
+                    },
+                };
+                if have_enter_scope {
+                    assert!(!entered_scope);
+                    result = quote! {
+                        #scope_enter
+                        #result
+                    };
+                    entered_scope = true;
+                }
+
+                if args_def.is_empty() {
+                    Some(result)
+                } else {
+                    // if we have args wrap the result in a block to prevent ident clashes.
+                    Some(quote! {{
+                        #(#args_def)*
+                        #result
+                    }})
+                }
+            })
+            .collect();
+
+        let body = if KIND_BLACK_LIST.contains(&ident.to_string().as_str()) {
+            let unused =
+                if fields_visits.is_empty() { Some(quote!(let _ = (visitor, it);)) } else { None };
+            quote! {
+                insert!("// TODO: AstKind doesn't exists");
+                #(#fields_visits)*
+                #unused
+            }
+        } else {
+            quote! {
+                let kind = AstKind::#ident(visitor.alloc(it));
+                visitor.enter_node(kind);
+                #(#fields_visits)*
+                visitor.leave_node(kind);
+            }
+        };
+
+        match (scope_enter, scope_leave, entered_scope) {
+            (_, Some(leave), true) => quote! {
+                #body
+                insert!("// TODO: cache this check!");
```

Comment:
@Boshen Sorry I forgot to cache this😓, Please don't merge accidentally before I mark it as ready again.